lightning-invoice/src/payment.rs

```diff
@@ -173,8 +187,45 @@ where
 	logger: L,
 	event_handler: E,
 	/// Caches the overall attempts at making a payment, which is updated prior to retrying.
-	payment_cache: Mutex<HashMap<PaymentHash, usize>>,
-	retry_attempts: RetryAttempts,
+	payment_cache: Mutex<HashMap<PaymentHash, PaymentAttempts<T>>>,
+	retry: Retry,
+}
+
+/// Storing minimal payment attempts information required for determining if a outbound payment can
+/// be retried.
+#[derive(Debug, Clone, Copy)]
+struct PaymentAttempts<T: Time> {
+	/// This should be >= 1 as we only insert PaymentAttempts for a PaymentHash if there is at least
+	/// /// one attempt.
+	counts: usize,
+	/// This field is only used when retry is [`Retry::Timeout`] which is only build with feature std
+	first_attempted_at: T
+}
+
+impl <T: Time>PaymentAttempts<T> {
+	fn new() -> Self {
+		PaymentAttempts{
+			counts: 1,
```

Comment:
I'm not sure why we insert at 1 here, that's a behavior change from the original code, which always inserted at 0.

Reply:
Yeah, also see: https://github.com/lightningdevkit/rust-lightning/pull/1418#discussion_r849708859

I'm not sure I follow why the count logic needs to change if you are just adding another strategy.

Reply:
My understanding of previous code is the count we store is the number of *retry*. 

For example, in `pay_pubkey`, it's set as 0 when there is no entry found in payment_cache.
https://github.com/lightningdevkit/rust-lightning/pull/1418/files#diff-5165a75ba263a549c7fb9f4b555e449e44b0b680a80a9715a99d11157802a0d7L330

However, in this PR, I changed it to store number of attempt(which should be number of *retry* + 1). And we will only want to create and store PaymentAttempts when there is at least one attempt. That's the reason why count is set as 1 in `PaymentAttempts::new`.

Reply:
I have actually tried to change storing total attempt to storing retry count(like current behaviour). https://github.com/lightningdevkit/rust-lightning/pull/1418/commits/145bede6055b25a91a8dba1596fd3ac8c991221b

It does look cleaner :). Please let me know if you still find it weird that I have to do `- 1` here https://github.com/lightningdevkit/rust-lightning/pull/1418/commits/145bede6055b25a91a8dba1596fd3ac8c991221b#diff-5165a75ba263a549c7fb9f4b555e449e44b0b680a80a9715a99d11157802a0d7R475

The biggest reason is I think `Retry::is_retryable_now` shouldn't be aware that `PaymentAttempts` passed has just been added 1 retry count. It should simply take `PaymentAttempts` and determine if it's retryable right now based on the current field. (My apology if my explanation is still confusing)

Reply:
Nice! Looks much better modulo Matt's comment about incrementing after calling `is_retryable_now`.